exchanges/kucoin/kucoin_websocket.go

```diff
@@ -1777,3 +1580,157 @@ func (ku *Kucoin) CalculateAssets(topic string, cp currency.Pair) ([]asset.Item,
+	case subscription.CandlesChannel:
+		return marketCandlesChannel // No support in GCT yet for Futures candles
+	case subscription.AllTradesChannel:
+		return marketMatchChannel // No support in GCT yet for Futures all trades
+	}
+	return s.Channel
+}
+
+// removeSpotFromMargin removes spot pairs from margin pairs in the supplied AssetPairs map for subscriptions to non-margin endpoints
+func removeSpotFromMargin(s *subscription.Subscription, ap map[asset.Item]currency.Pairs, spotPairs currency.Pairs) string {
+	if strings.HasPrefix(s.Channel, "/margin") {
+		return ""
+	}
+	if p, ok := ap[asset.Margin]; ok {
+		ap[asset.Margin] = p.Remove(spotPairs...)
+	}
+	return ""
+}
+
+// isSymbolChannel returns if the channel expects receive a symbol
+func isSymbolChannel(s *subscription.Subscription) bool {
+	switch channelName(s, s.Asset) {
+	case privateSpotTradeOrders, accountBalanceChannel, marginPositionChannel, spotMarketAdvancedChannel, futuresSystemAnnouncementChannel,
+		futuresTradeOrderChannel, futuresStopOrdersLifecycleEventChannel, futuresAccountBalanceEventChannel:
+		return false
+	}
+	return true
+}
+
+// isCurrencyChannel returns if the channel expects receive a currency
+func isCurrencyChannel(s *subscription.Subscription) bool {
+	return s.Channel == marginLoanChannel
+}
+
+// channelInterval returns the channel interval if it has one
+func channelInterval(s *subscription.Subscription) string {
+	if channelName(s, s.Asset) == marketCandlesChannel {
+		if i, err := intervalToString(s.Interval); err == nil {
+			return i
+		}
+	}
+	return ""
+}
+
+// assetCurrencies returns the currencies from all pairs in an asset
+// Updates the AssetPairs map parameter to contain only those currencies as Base items for expandTemplates to see
+func assetCurrencies(s *subscription.Subscription, ap map[asset.Item]currency.Pairs) currency.Currencies {
+	cs := common.SortStrings(ap[s.Asset].GetCurrencies())
+	p := currency.Pairs{}
+	for _, c := range cs {
+		p = append(p, currency.Pair{Base: c})
+	}
+	ap[s.Asset] = p
+	return cs
+}
+
+// joinPairsWithInterval returns a list of currency pair symbols joined by comma
+// If the subscription has a viable interval it's appended after each symbol
+func joinPairsWithInterval(b currency.Pairs, s *subscription.Subscription) string {
+	out := make([]string, len(b))
+	suffix, err := intervalToString(s.Interval)
+	if err == nil {
+		suffix = "_" + suffix
+	}
+	for i, p := range b {
+		out[i] = p.String() + suffix
+	}
+	return strings.Join(out, ",")
+}
+
+const subTplText = `
+{{- removeSpotFromMargin $.S $.AssetPairs -}}
+{{- if isCurrencyChannel $.S }}
+	{{ channelName $.S $.S.Asset -}} : {{- (assetCurrencies $.S $.AssetPairs).Join -}}
+{{- else if isSymbolChannel $.S }}
+	{{ range $asset, $pairs := $.AssetPairs }}
+		{{- with $name := channelName $.S $asset }}
+			{{- if and (eq $name "/market/ticker") (gt (len $pairs) 10) -}}
+				{{- $name -}} :all
```

Comment:
I know you hate me at this point but I have to test things 😅 
Here's this:
```
[ERROR] | WEBSOCKET | 07/08/2024 10:16:01 | Kucoin websocket: subscription failure, ticker all : subscription template did not generate the expected number of pair records for spot: Got 1; Expected 17
```
Here's some config snippets:
```json
 "futures": {
      "assetEnabled": true,
      "enabled": "XBT_USDTM,ETH_USDTM",
```

```json
  "margin": {
      "assetEnabled": true,
      "enabled": "BTC-USDT",
```

```json
 "spot": {
      "assetEnabled": true,
      "enabled": "AVA-USDT,FET-BTC,FET-ETH,ANKR-BTC,XMR-BTC,XMR-ETH,MTV-BTC,MTV-ETH,CRO-BTC,MTV-USDT,KMD-BTC,KMD-USDT,TEL-USDT,TT-USDT,AERGO-USDT,XMR-USDT,TRX-KCS",
```

only enabled sub:
```json
 {
      "enabled": true,
      "channel": "ticker",
```

Reply:
Fixed cfb62def

Lost the batchSize for :all at some point.

Reply:
Also added a test for this:

Fixed 7e118b29

Reply:
Okay, so this was a real edge case.
xpandPairs was set to true because futures and margin needed to expandPairs.
So testing this with asset: spot would have a different result from asset: All

The *second* correct fix for this was to also move the xpandPairs down into AssetPairs

Fixed 51ddad04 and 65ee7043